### ISSUE TITLE
[TF 2.0] Different results of binary cross entropy loss of the same architecture 

### ISSUE THREAD

**Issue by jacekblaz** [tensorflow/tensorflow#31930]
**System information**
- Tensorflow 2.0

**Describe the current behavior**
I have created model from tf.keras.models for binary classification problem. It is dense with sigmoid activation in the last layer and binary cross entropy loss. I got different loss values whether my last layer was:
```last_layer = Dense(1, activation='sigmoid')(previous_layer)```
or
```
last_layer = Dense(1)(previous_layer)
last_layer = sigmoid(last_layer)
```

If `loss=BinaryCrossentropy(from_logits=False)`:
-In the first case BCE do apply sigmoid before computing loss.
-In the second case BCE do not apply sigmoid before computing loss.

**Describe the expected behavior**
I believe that from user point of view it would consistent if both examples gives exactly the same output- there is no difference in architecture.

It would be better if sigmoid is applied, no matter what are the initialization arguments of BinaryCrossentropy.

**Code to reproduce the issue**
```
from tensorflow.python.keras.layers import Input, Dense
from tensorflow.python.keras.activations import sigmoid
from tensorflow.python.keras.models import Model
from tensorflow.python.keras.losses import BinaryCrossentropy
import numpy as np


x_train = np.array([[1000]])
y_train = [0]

inp = Input(x_train.shape[1])
out = Dense(1, trainable=False, use_bias=False)(inp)
model = Model(inp, out)
model.get_layer('dense').set_weights([np.array([[1]])])
model.compile(optimizer='adam', loss=BinaryCrossentropy(from_logits=True))

print('\nFrom logits=True, no activation')
model.fit(x_train, y_train, epochs=1)
print('Prediction:')
print(model.predict(x_train))


inp = Input(x_train.shape[1])
out = Dense(1, trainable=False, use_bias=False)(inp)
out = sigmoid(out)
model = Model(inp, out)
model.get_layer('dense_1').set_weights([np.array([[1]])])
model.compile(optimizer='adam', loss=BinaryCrossentropy(from_logits=False))

print('\nFrom logits=False, separate sigmoid')
model.fit(x_train, y_train, epochs=1)
print('Prediction:')
print(model.predict(x_train))


inp = Input(x_train.shape[1])
out = Dense(1, trainable=False, use_bias=False, activation='sigmoid')(inp)
model = Model(inp, out)
model.get_layer('dense_2').set_weights([np.array([[1]])])
model.compile(optimizer='adam', loss=BinaryCrossentropy(from_logits=False))

print('\nFrom logits=False, sigmoid in dense')
model.fit(x_train, y_train, epochs=1)
print('Prediction:')
print(model.predict(x_train))
```

Output tf 2.0:
```
From logits=True, no activation
1/1 [==============================] - 0s 17ms/sample - loss: 1000.0000
Prediction:
[[1000.]]

From logits=False, separate sigmoid
Train on 1 samples
1/1 [==============================] - 0s 15ms/sample - loss: 1000.0000
Prediction:
[[1.]]

From logits=False, sigmoid in dense
Train on 1 samples
1/1 [==============================] - 0s 17ms/sample - loss: 15.3332
Prediction:
[[1.]]
```

output tf 1.14

```
From logits=True, no activation
1/1 [==============================] - 0s 39ms/sample - loss: 1000.0000
Prediction:
[[1000.]]

From logits=False, separate sigmoid
1/1 [==============================] - 0s 44ms/sample - loss: 1000.0000
Prediction:
[[1.]]

From logits=False, sigmoid in dense
1/1 [==============================] - 0s 57ms/sample - loss: 1000.0000
Prediction:
[[1.]]
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gowthamkpr** [comment]
Can you please provide the github gist of the code for TF1.14  @jacekblaz

**Comment by jacekblaz** [comment]
@gowthamkpr 
[https://colab.research.google.com/gist/jacekblaz/1b860965969dca734a7618c2009bf018/untitled.ipynb](gist)

**Comment by pavithrasv** [comment]
Thank you for the issue, i will look into this. I believe that some change was made when `from_logits=False` and last layer is `sigmoid` to use the `logits` from the previous layer. I am suspecting there is some issue there.

**Comment by pavithrasv** [comment]
I could not repro this on the latest nightly, are you still seeing this?

**Comment by jacekblaz** [comment]
@gowthamkpr I believe you are encountering the issue in TF 2.0b1- there is different output for the same architecture in your gist. This issue is not about error ValueError caused by numpy. 

@pavithrasv I have tried to reproduce this error in current nightly as well as in 2.0.0-rc1 and in both it seems to be fixed- results are the same: 1000.0

[2.0.0-rc1](https://colab.research.google.com/gist/jacekblaz/1b860965969dca734a7618c2009bf018/untitled.ipynb)

[2.0-b1](https://colab.research.google.com/drive/1Rtc84ogFactv-gE89A2QXV-6pwDKlybg)

[nightly](https://colab.research.google.com/drive/1BPrwX-ffEqjOzi-WrML-wCcfBftyf5aI)

So at least the same architecture written in different style provide the same results. However I am not sure if that result is correct. Logits has values between 0 and 1, maximum BinaryCrossentropy for those value is 15.424949. I believe BinaryCrossentropy(from_logits=False) should compute logits before applying BinaryCrossentropy itself. Therefore if input=1000, target=0 and BinaryCrossentropy(from_logits=False) the result should be max 15.424949. 

I think documentation, code, behavior and math behind it are inconsistent. Look at this [comment](https://github.com/tensorflow/tensorflow/blob/c6babdd8aae6d71d0b570216557d20983d6421e7/tensorflow/python/keras/backend.py#L4470) deeply hidden in backend

**Comment by yarin05** [comment]
This one deserve more attention guys, looks like a bug to me @dynamicwebpaige @pavithrasv and @karmel  
I am unable to reproduce my results with image segmentation model when changing only the training loop style to work with tf.function 
working style:
    model.compile(optimizer=optimizer, loss=['categorical_crossentropy'])
    model.train_on_batch(inputs, outputs])   
broken style:
    categorical_loss = tf.keras.losses.CategoricalCrossentropy()
    train_step() # which is tf.function
in my case the loss of the broken style start from ~15.385 and in the working style from ~5

**Comment by pavithrasv** [comment]
@jacekblaz I found out what's happening here. We have two different functions - one that computes cross-entropy loss when we have the logits information, one that computes cross-entropy loss with probabilities. For the inputs you have provided label=0 and predictions=1000, the loss value is undefined. Each of these functions handle this undefined case differently and hence you see 1000 for one and 15.33 for the other.

**Comment by pavithrasv** [comment]
We recommend that you use logits as inputs always for numerical stability. All of the TF docs and examples will be updated to reflect the same.
